app/models/miq_server.rb

```diff
@@ -346,6 +351,15 @@ def monitor
     end
   end
 
+  def monitor_myself
+    if memory_usage.to_i > memory_threshold
+      msg = "server(pid: #{pid}, name: #{name}) memory usage [#{memory_usage.to_i}] exceeded limit: [#{memory_threshold}].  Exiting server process."
+      _log.warn(msg)
+      MiqEvent.raise_evm_event_queue_in_region(self, "evm_server_memory_exceeded", :event_details => msg, :type => self.class.name)
+      exit 1
```

Comment:
Does it make sense to `raise SystemExit` instead here, so it is caught in the above `rescue`, and we do the proper shutdown procedure in the `monitor_loop` method?  Or would calling `shutdown`/`shudown_and_exit` make more sense?

Reply:
Good questions @NickLaMuro.  So, it's important to exit with a non-zero exit code.  We can do `shutdown_and_exit(1)` which will raise another event (stop), quiesce workers and exit with 1.

Doing a exit 1 here doesn't quiesce workers, so they're expected to exit on their own when they try to hearbeat to a non-existing drb server before they're killed by the replacement server that starts a few minutes later via systemd/kubenetes.

I can be convinced to use `shutdown_and_exit(1)`.

Reply:
I'll test `shutdown_and_exit(1)` and report the results

Reply:
Ok, so the main difference appears that we deactivate all roles on the server, quiesce workers (possibly waiting up to 5 minutes (IIRC)), clean the worker rows, clean any worker queue messages, etc.  This seems to be better:


```
[----] W, [2018-01-11T13:40:59.095435 #8345:b8f13c]  WARN -- : MIQ(MiqServer#monitor_myself) server(pid: 8345, name: EVM) memory usage [354418688] exceeded limit: [346030080].  Exiting server process.
[----] I, [2018-01-11T13:40:59.099947 #8345:b8f13c]  INFO -- : MIQ(MiqQueue.put) Message id: [9893],  id: [], Zone: [], Role: [], Server: [], Ident: [generic], Target id: [], Instance id: [], Task id: [], Command: [MiqEvent.raise_evm_event], Timeout: [600], Priority: [100], State: [ready], Deliver On: [], Data: [], Args: [["MiqServer", 1], "evm_server_memory_exceeded", {:event_details=>"server(pid: 8345, name: EVM) memory usage [354418688] exceeded limit: [346030080].  Exiting server process.", :type=>"MiqServer"}]
[----] I, [2018-01-11T13:40:59.100195 #8345:b8f13c]  INFO -- : MIQ(MiqServer#shutdown) initiated for MiqServer [EVM] with ID: [1], PID: [8345], GUID: [6d28a0a7-3b2b-4da4-9306-6799c74a4295]
[----] I, [2018-01-11T13:40:59.116440 #8345:b8f13c]  INFO -- : MIQ(MiqQueue.put) Message id: [9894],  id: [], Zone: [default], Role: [automate], Server: [], Ident: [generic], Target id: [], Instance id: [], Task id: [], Command: [MiqAeEngine.deliver], Timeout: [3600], Priority: [20], State: [ready], Deliver On: [], Data: [], Args: [{:object_type=>"MiqServer", :object_id=>1, :attrs=>{:event_type=>"evm_server_stop", "MiqEvent::miq_event"=>458, :miq_event_id=>458, "EventStream::event_stream"=>458, :event_stream_id=>458}, :instance_name=>"Event", :user_id=>1, :miq_group_id=>2, :tenant_id=>1, :automate_message=>nil}]
[----] I, [2018-01-11T13:40:59.126574 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <ems_inventory> on Server <EVM>
[----] I, [2018-01-11T13:40:59.130993 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <ems_operations> on Server <EVM>
[----] I, [2018-01-11T13:40:59.135073 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <event> on Server <EVM>
[----] I, [2018-01-11T13:40:59.138552 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <smartstate> on Server <EVM>
[----] I, [2018-01-11T13:40:59.142318 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <automate> on Server <EVM>
[----] I, [2018-01-11T13:40:59.145567 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <database_operations> on Server <EVM>
[----] I, [2018-01-11T13:40:59.149075 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <reporting> on Server <EVM>
[----] I, [2018-01-11T13:40:59.152524 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <scheduler> on Server <EVM>
[----] I, [2018-01-11T13:40:59.156186 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <user_interface> on Server <EVM>
[----] I, [2018-01-11T13:40:59.159255 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <websocket> on Server <EVM>
[----] I, [2018-01-11T13:40:59.161987 #8345:b8f13c]  INFO -- : MIQ(AssignedServerRole#deactivate) Deactivating Role <web_services> on Server <EVM>
[----] I, [2018-01-11T13:40:59.163744 #8345:b8f13c]  INFO -- : MIQ(MiqServer#quiesce_workers_loop) Stopping all active workers
```

```
[----] I, [2018-01-11T13:40:59.166872 #8345:b8f13c]  INFO -- : MIQ(MiqServer#stop_worker) Stopping Worker [MiqEventHandler] with ID: [324], PID: [8624], GUID: [eb10b440-f2ee-4c83-a592-feb738ce1b43], status [started]...
[----] I, [2018-01-11T13:40:59.171316 #8345:b8f13c]  INFO -- : MIQ(MiqQueue.put) Message id: [9895],  id: [], Zone: [default], Role: [], Server: [], Ident: [generic], Target id: [], Instance id: [], Task id: [], Command: [MiqEvent.raise_evm_event], Timeout: [600], Priority: [100], State: [ready], Deliver On: [], Data: [], Args: [["MiqServer", 1], "evm_worker_stop", {:event_details=>"Stopping Worker [MiqEventHandler] with ID: [324], PID: [8624], GUID: [eb10b440-f2ee-4c83-a592-feb738ce1b43], status [started]...", :type=>"MiqEventHandler"}]
[----] I, [2018-01-11T13:40:59.174237 #8345:b8f13c]  INFO -- : MIQ(MiqServer#worker_set_message) Worker [MiqEventHandler] with ID: [324], PID: [8624], GUID: [eb10b440-f2ee-4c83-a592-feb738ce1b43] is being requested to exit
```

```
[----] I, [2018-01-11T13:41:02.285512 #8345:b8f13c]  INFO -- : MIQ(MiqServer#clean_worker_records) SQL Record for Worker [MiqEventHandler] with ID: [324], PID: [8624], GUID: [eb10b440-f2ee-4c83-a592-feb738ce1b43], Status: [stopped] is being deleted
```

```
[----] I, [2018-01-11T13:41:06.464230 #8345:b8f13c]  INFO -- : MIQ(MiqServer#quiesce_all_workers) Cleaning all active messages being processed by MiqServer
```

Reply:
@gtanzillo @dmetzger57 @NickLaMuro @Fryguy if we deactivate roles on this server on this server, it's likely they'll be migrated to another appliance so the new server process that gets started may have different roles.  If we just exit 1, we might start the new server before the master server reassigns roles.  Which way do you think is better?

Reply:
I think there can never be enough Nicks... cc @carbonin since we were chatting about shutdown_and_exit the other day

Reply:
Assuming what you say is true (I don't know myself), then maybe what you have already makes sense.  That said, maybe this is something we test in a multi-server environment?

Reply:
> we might start the new server before the master server reassigns roles

I'm not sure we should strive for this with this change. I would rather shutdown cleanly in this case (`shutdown_and_exit(1)`) and make sure things are okay for the new server process to start.

I think exiting is something we don't do now too often so I'm not as comfortable with relying on the workers erroring out like you mentioned, I would rather us tell them to come down.